mtags/src/main/scala-2/scala/meta/internal/pc/InferredTypeProvider.scala

```diff
@@ -0,0 +1,185 @@
+      /* `val a = 1` or `var b = 2`
+       *     turns into
+       * `val a: Int = 1` or `var b: Int = 2`
+       */
+      case vl @ ValDef(_, name, tpt, _) if !vl.symbol.isParameter =>
+        // it seems name might sometime contain the space afterwards
+        val nameEnd = tpt.pos.start + name.toString.trim().size
+        val nameEndPos = tpt.pos.withEnd(nameEnd).withStart(nameEnd).toLSP
+        val typeNameEdit = new TextEdit(nameEndPos, ": " + prettyType(tpt.tpe))
+        typeNameEdit :: additionalImports
+
+      /* `.map(a => a + a)`
+       *     turns into
+       * `.map((a: Int) => a + a)`
+       */
+      case vl @ ValDef(_, name, tpt, _) if vl.symbol.isParameter =>
+        val nameEnd = vl.pos.start + name.length()
+        val namePos = tpt.pos.withEnd(nameEnd).withStart(nameEnd).toLSP
+        def typeNameEdit(rightParen: String) =
+          new TextEdit(namePos, ": " + prettyType(tpt.tpe) + rightParen)
+
+        def paramStart = vl.pos.withEnd(vl.pos.start).toLSP
+        def startParamEdit = new TextEdit(paramStart, "(")
+
+        val needsParens = lastVisitedParentTrees match {
+          /* Find how the function starts either with:
+           * - `(` - then we need braces
+           * - `{` - then we don't need braces
+           */
+          case _ :: (_: Function) :: (appl: Apply) :: _ =>
+            params
+              .text()
+              .substring(appl.pos.start, vl.pos.start)
+              .tokenize
+              .toOption
+              .exists {
+                _.tokens.reverseIterator
+                  .find(t => t.is[T.LeftParen] || t.is[T.LeftBrace])
+                  .exists(_.is[T.LeftParen])
+              }
+          case _ => false
+        }
+
+        if (needsParens) {
+          startParamEdit :: typeNameEdit(")") :: additionalImports
+
+        } else {
+          typeNameEdit("") :: additionalImports
+        }
+
+      /* `def a[T](param : Int) = param`
+       *     turns into
+       * `def a[T](param : Int): Int = param`
+       */
+      case DefDef(_, name, _, _, tpt, rhs) =>
+        val nameEnd = tpt.pos.start + name.length()
+
+        // search for `)` or `]` or defaut to name's end to insert type
+        val searchString = params
+          .text()
+          .substring(nameEnd, rhs.pos.start) // cotains the parameters and =
+        val lastTokenPos = searchString.tokenize.toOption
+          .flatMap(tokens =>
+            tokens.tokens.reverseIterator
+              .find(t => t.is[T.RightParen] || t.is[T.RightBracket])
+              .map(t => nameEnd + t.pos.end)
+          )
+          .getOrElse(nameEnd)
+
+        val insertPos = rhs.pos.withStart(lastTokenPos).withEnd(lastTokenPos)
+        val typeNameEdit =
+          new TextEdit(insertPos.toLSP, ": " + prettyType(tpt.tpe))
+
+        typeNameEdit :: additionalImports
+
+      /* `case t =>`
+       *  turns into
+       * `case t: Int =>`
+       */
+      case bind @ Bind(name, body) =>
```

Comment:
is this a safe/desired refactor? 

Turning 

```scala
def a: Int = ???
a match {
  case n if n > 2 => ??? 
  case n          => ???
}
```

into

```scala
def a: Int = ???
a match {
  case n if n > 2 => ??? 
  case n: Int     => ???
}
```

it's not a type annotation, but it's actually inserting a runtime check. I'm not sure I would put it in the same basket as the other refactors this PR covers.

Reply:
I think that should be safe. It's the actual type of a, so the check is basically checking if Int is Int. If anyone changes the type of a the compiler will issue a warning.

We could name it differently similar to how Intellij does it `add type annotation to pattern definition`.

Reply:
👍 on naming it like that

Reply:
Changed!